File: nltemplate.f90
function cfun(label,x,y,z,ur,dur,nvar,istep, &
          ireg,iregup,rnode,vec,imax,jmax,kmax)

   character(*) label
   integer nvar,istep,imax,jmax,kmax
   double precision cfun,x,y,z,ur(nvar),dur(nvar,3)
   integer ireg(0:imax,0:jmax,0:kmax),iregup(0:imax,0:jmax,0:kmax)
   double precision rnode(0:imax,0:jmax,0:kmax,3),vec(*)

! Your code goes here

end function cfun

! ######################################################################
     
function afun(label,x,y,z,ur,dur,nvar,istep, &
          ireg,iregup,rnode,vec,imax,jmax,kmax)

   character(*) label
   integer nvar,istep,imax,jmax,kmax
   double precision afun,x,y,z,ur(nvar),dur(nvar,3)
   integer ireg(0:imax,0:jmax,0:kmax),iregup(0:imax,0:jmax,0:kmax)
   double precision rnode(0:imax,0:jmax,0:kmax,3),vec(*)

! Your code goes here

end function afun

! ######################################################################

function ffun(label,x,y,z,ur,dur,nvar,istep, &
          ireg,iregup,rnode,vec,imax,jmax,kmax)

   character(*) label  
   integer nvar,istep,imax,jmax,kmax
   double precision x,y,z,ffun,ur(nvar),dur(nvar,3)
   integer ireg(0:imax,0:jmax,0:kmax),iregup(0:imax,0:jmax,0:kmax)
   double precision rnode(0:imax,0:jmax,0:kmax,3),vec(*)

! Your code goes here

end function ffun

! ######################################################################

function qfun(label,x,y,z,nx,ny,nz,ur,nvar,istep, &
          ireg,iregup,rnode,vec,imax,jmax,kmax)

   character(*) label
   double precision qfun,x,y,z,nx,ny,nz,ur(nvar)
   integer nvar,istep,imax,jmax,kmax
   integer ireg(0:imax,0:jmax,0:kmax),iregup(0:imax,0:jmax,0:kmax)
   double precision rnode(0:imax,0:jmax,0:kmax,3),vec(*)

! Your code goes here

end function qfun

! ######################################################################
     
function gfun(label,x,y,z,nx,ny,nz,ur,nvar,istep, &
          ireg,iregup,rnode,vec,imax,jmax,kmax)

   character(*) label
   double precision gfun,x,y,z,nx,ny,nz,ur(nvar)
   integer nvar,istep,imax,jmax,kmax
   integer ireg(0:imax,0:jmax,0:kmax),iregup(0:imax,0:jmax,0:kmax)
   double precision rnode(0:imax,0:jmax,0:kmax,3),vec(*)

! Your code goes here

end function gfun

! ######################################################################

function BCfun(label,x,y,z,ur,nvar,istep, &
          ireg,iregup,rnode,vec,imax,jmax,kmax)

   character(*) label
   double precision BCfun,x,y,z,ur(nvar)
   integer nvar,istep,imax,jmax,kmax
   integer ireg(0:imax,0:jmax,0:kmax),iregup(0:imax,0:jmax,0:kmax)
   double precision rnode(0:imax,0:jmax,0:kmax,3),vec(*)

! Your code goes here

end function BCfun

! ######################################################################

function scanfun(label,x,y,z,nx,ny,nz,ur,dur,nvar)

  character(*) label
  integer nvar
  double precision scanfun,x,y,z,nx,ny,nz,ur(nvar),dur(nvar,3)

! Your code goes here

end function scanfun

File: nltemplate.c
// Note: this template assumes you have 1 solved-for variable. Hence ur[1], dur[3][1]
// If you have N solved-for variables, you need to change to ur[N], dur[3][N]

double cfun(char *label, double *x, double *y, double *z, 
       double ur[1], double dur[3][1], int *nvar, int *istep, 
       int *ireg, int *iregup, double *rnode, double *vec, 
       int *imax, int *jmax, int *kmax, int length)
{
  // your code goes here
}

// ######################################################################

double afun(char *label, double *x, double *y, double *z, 
       double ur[1], double dur[3][1], int *nvar, int *istep, 
       int *ireg, int *iregup, double *rnode, double *vec, 
       int *imax, int *jmax, int *kmax, int length)
{
  // your code goes here
}


// ######################################################################

double ffun(char *label, double *x, double *y, double *z, 
       double ur[1], double dur[3][1], int *nvar, int *istep, 
       int *ireg, int *iregup, double *rnode, double *vec, 
       int *imax, int *jmax, int *kmax, int length)
{
  // your code goes here
}


// ######################################################################

double qfun(char *label, double  *x, double  *y, double  *z, 
                         double *nx, double *ny, double *nz,
       double ur[1], int *nvar, int *istep, 
       int *ireg, int *iregup, double *rnode, double *vec, 
       int *imax, int *jmax, int *kmax, int length)
{
  // your code goes here
}


// ######################################################################

double gfun(char *label, double  *x, double  *y, double  *z, 
                         double *nx, double *ny, double *nz,
       double ur[1], int *nvar, int *istep, 
       int *ireg, int *iregup, double *rnode, double *vec, 
       int *imax, int *jmax, int *kmax, int length)
{
  // your code goes here
}


// ######################################################################

double BCfun(char *label, double *x, double *y, double *z,
       double ur[1], int *nvar, int *istep, 
       int *ireg, int *iregup, double *rnode, double *vec, 
       int *imax, int *jmax, int *kmax, int length)
{
  // your code goes here
}


// ######################################################################

double scanfun(char *label, double  *x, double  *y, double  *z,
                            double *nx, double *ny, double *nz,
       double ur[1], int *nvar, int length)
{
  // your code goes here
}

